include/libtorrent/ssl/context.hpp

```diff
@@ -0,0 +1,279 @@
+    contributors may be used to endorse or promote products derived
+    from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_SSL_SSL_CONTEXT_HPP_INCLUDED
+#define TORRENT_SSL_SSL_CONTEXT_HPP_INCLUDED
+
+#include "libtorrent/config.hpp"
+
+#ifdef TORRENT_USE_SSL
+
+#include "libtorrent/aux_/disable_warnings_push.hpp"
+
+#include <boost/asio.hpp>
+#include <boost/asio/buffer.hpp>
+#include <boost/system/error_code.hpp>
+
+#ifndef BOOST_NO_EXCEPTIONS
+#include <boost/system/system_error.hpp>
+#endif
+
+#ifdef TORRENT_USE_OPENSSL
+#include <openssl/opensslv.h> // for OPENSSL_VERSION_NUMBER
+#if OPENSSL_VERSION_NUMBER < 0x1000000fL
+#error OpenSSL too old, use a recent version with SNI support
+#endif
+#ifdef TORRENT_WINDOWS
+// because openssl includes winsock.h, we must include winsock2.h first
+#include <winsock2.h>
+#endif
+#include <openssl/ssl.h>
+#include <openssl/x509v3.h> // for GENERAL_NAME
+#include <openssl/crypto.h>
+#include <openssl/rand.h>
+#include <boost/asio/ssl.hpp>
+#include <boost/asio/ssl/rfc2818_verification.hpp>
+#endif
+
+#ifdef TORRENT_USE_GNUTLS
+#include <gnutls/gnutls.h>
+#include <boost/asio/gnutls.hpp>
+#endif
+
+#include "libtorrent/aux_/disable_warnings_pop.hpp"
+
+#include <string>
+#include <cstddef>
+#include <functional>
+#include <exception>
+
+namespace libtorrent {
+namespace ssl {
+
+namespace error {
+
+#if defined TORRENT_USE_OPENSSL
+using boost::asio::error::get_ssl_category;
+#if BOOST_VERSION >= 106400
+using boost::asio::ssl::error::get_stream_category;
+#else
+static const boost::system::error_category& get_stream_category() { return get_ssl_category(); }
+#endif
+#elif defined TORRENT_USE_GNUTLS
+using boost::asio::gnutls::error::get_ssl_category;
+using boost::asio::gnutls::error::get_stream_category;
+#endif
+
+}
+
+using error_code = boost::system::error_code;
+using const_buffer = boost::asio::const_buffer;
+
+#if defined TORRENT_USE_OPENSSL
+namespace base = boost::asio::ssl;
+class stream_base;
+#elif defined TORRENT_USE_GNUTLS
+namespace base = boost::asio::gnutls;
+using base::stream_base;
+#endif
+
+class TORRENT_EXTRA_EXPORT context : public base::context
```

Comment:
It would seem more appropriate to just make `context` a typedef to either the boost asio one or the gnutls one. As far as I can tell, all OpenSSL/GnuTLS specific code could still be done in free functions (ideally implemented in a .cpp file)

Reply:
A significant amount of code can be moved to .cpp files for sure. 
Free functions would have a drawback I think: I see no way to implement generic functions for `set_servername_callback` (because you need to keep the generic callback somewhere and openssl takes a raw function pointer) and `get_context` because you need to retrieve the wrapped context somehow).

Reply:
but boost.asio's `ssl::context` already keeps that function object. You just have to do the same thing in the gnutls counterpart, no?

Reply:
It does for `set_verify_callback`, but it has no support for servername callback sadly.